File: 11. 盛最多水的容器.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
    let left = 0;
    let right = height.length - 1;

    let maxCapacity = 0;
    while (left < right) {
        const capacity = (right - left) * Math.min(height[left], height[right]);
        maxCapacity = Math.max(maxCapacity, capacity);

        if (height[left] < height[right]) {
            left++;
        } else {
            right--;
        }
    }

    return maxCapacity;
};
//  var maxArea = function(height) {
//     var maxCapacity = 0;
//     var st = 0;
//     var ed = height.length - 1;
//     while (st < ed) {
//         maxCapacity = Math.max(maxCapacity, Math.min(height[st], height[ed]) * (ed - st));
//         if (height[st] < height[ed]) {
//             var nowMinIndex = st;
//             while (st < ed) {
//                 st++;
//                 if (height[st] > height[nowMinIndex]) break;
//             }
//         } else {
//             var nowMinIndex = ed;
//             while (st < ed) {
//                 ed--;
//                 if (height[ed] > height[nowMinIndex]) break;
//             }
//         }
//     }
//     return maxCapacity;
// };

File: 11. 盛最多水的容器.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int left = 0;
        int right = height.size() - 1;

        int maxCapacity = 0;
        while (left < right) {
            int capacity = (right - left) * min(height[left], height[right]);
            maxCapacity = max(maxCapacity, capacity);

            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }

        return maxCapacity;
    }
};